# Notebook Instructions

1. If you are new to Jupyter notebooks, please go through this introductory manual <a href='https://quantra.quantinsti.com/quantra-notebook' target="_blank">here</a>.
1. Any changes made in this notebook would be lost after you close the browser window. **You can download the notebook to save your work on your PC.**
1. Before running this notebook on your local PC:<br>
i.  You need to set up a Python environment and the relevant packages on your local PC. To do so, go through the section on "**Run Codes Locally on Your Machine**" in the course.<br>
ii. You need to **download the zip file available in the last unit** of this course. The zip file contains the data files and/or python modules that might be required to run this notebook.

# Implementation of Kelly Criterion

Kelly criterion `K%` formula is used by traders and investors to manage their trade size. The equation uses:
1. Winning probability (W) as the expectation of a trade being positive
2. Win/Loss ratio (R)

After completing this notebook, you will be able to implement the Kelly criterion in Python.

In this notebook, you will perform the following steps:

1. [Read the Data](#read)
2. [Kelly Criterion Logic](#strategy_logic)

## Import Libraries

In [1]:
# Data manipulation
import pandas as pd

# Ignore warnings
import warnings 
warnings.filterwarnings('ignore')

# Set path
import sys
sys.path.append("..")

<a id='read'></a>
## Read the Data
The index reversal strategy's trading signal and the portfolio value are read from a file named `strategy_data.csv`.

In [2]:
# Updating the dataframe to keep only trade day
strategy = pd.read_csv('../data_modules/strategy_data.csv',
                       index_col=0, parse_dates=True)

# Preview the strategy data
strategy.head()

signal   returns  cumulative_returns
2005-01-03       0 -0.000000            1.000000
2005-01-04       1 -0.000000            1.000000
2005-01-05       1 -0.003537            0.996463
2005-01-06       0  0.003296            0.999747
2005-01-07       0 -0.000000            0.999747

<a id='strategy_logic'></a>
## Kelly Criterion Logic
1. Count number of winning and losing trades.
2. Calculate `Average Win` and `Average Loss` per trade.
3. Calculate winning probability `W` and Win/Loss ratio `R`.
4. Calculate `K%`.

#### Count the number of winning and losing trades
Since strategy is being rebalanced everyday, to count the total number of trades, we will calculate the total number of non-zero returns.

In [3]:
# Count number of trades
number_of_trades = strategy['signal'].sum()
print("Number of trades: ", number_of_trades)

# Count the number of positive trades
positive_trades = strategy[strategy['returns'] > 0].count()[0]
print("Number of positive trades: ", positive_trades)

# Count the number of negative trades
negative_trades = strategy[strategy['returns'] < 0].count()[0]
print("Number of negative trades: ", negative_trades)

Number of trades:  503
Number of positive trades:  297
Number of negative trades:  206


#### Calculate `Average Win` and `Average Loss` per trade

In [4]:
# Calculate average return for winning trades
average_win = (strategy['returns'][strategy['returns'] > 0].mean())

# Calculate average return for losing trade
average_loss = (strategy['returns'][strategy['returns'] < 0].mean())

print("Average win amount: ", round(average_win, 2))
print("Average loss amount: ", round(abs(average_loss), 2))

Average win amount:  0.01
Average loss amount:  0.01


#### Calculate Winning probability and Win/Loss ratio
$$\text{Win probability = }\frac{Number\ of\ positive\ trades}{Total\ \ number \ of\ trades}$$


$$\text{Win/Loss ratio = }\frac{Average\ winning\ trade\ amount}{Average\ losing\ trade\ amount}$$

In [5]:
# Calculate winning probability
W = positive_trades / number_of_trades

# Calculate Win/Loss ratio
R = average_win / abs(average_loss)

#### Calculate Kelly Criterion

**Formula for Kelly Criterion**: 
$$ K\% = W - \frac{(1 - W)}{R}$$

In [6]:
# Calculate position size from Kelly criterion
K = W - (1 - W)/R

print("Position size as per Kelly criterion: ", round(K*100, 2), "%")

Position size as per Kelly criterion:  20.32 %


Therefore, we should use 20.32% of our account value as our trade size.

## Tweak the Code
You can tweak the code in the following ways:

1. Use returns from a different strategy or a list of trades to find the optimal Kelly.
2. Change the rules of the current trading strategy to see the effect on K.
3. Find optimum Kelly using a subset of the returns. Use the value obtained to backtest the strategy with rest of the returns. 

In the upcoming unit, we will discuss the limitations of Kelly criterion.<br><br>